jp2/tests/parse_tests.rs

```diff
@@ -0,0 +1,231 @@
+    };
+    test_jp2_file(filename, expected_configuration);
+}
+
+fn test_jp2_file(filename: &str, expected: ExpectedConfiguration) {
+    let path = Path::new(env!("CARGO_MANIFEST_DIR"))
+        .join("tests")
+        .join(filename);
+    let file = File::open(path).expect("file should exist");
+    let mut reader = BufReader::new(file);
+    let result = decode_jp2(&mut reader);
+    assert!(result.is_ok());
+    let boxes = result.unwrap();
+    assert_eq!(boxes.length(), expected.file_length);
+
+    assert!(boxes.signature_box().is_some());
+    let signature = boxes.signature_box().as_ref().unwrap();
+    assert_eq!(signature.signature(), *b"\x0d\x0a\x87\x0a");
+
+    assert!(boxes.file_type_box().is_some());
+    let file_type = boxes.file_type_box().as_ref().unwrap();
+    assert_eq!(file_type.brand(), "jp2 ");
+    assert_eq!(file_type.min_version(), 0);
+    assert_eq!(file_type.compatibility_list(), vec!["jp2 "]);
+
+    assert!(boxes.header_box().is_some());
+    let header_box = boxes.header_box().as_ref().unwrap();
+    let image_header_box = &header_box.image_header_box;
+    assert_eq!(image_header_box.height(), 128);
+    assert_eq!(image_header_box.width(), 64);
+    assert_eq!(image_header_box.components_num(), 3);
+    assert_eq!(image_header_box.compression_type(), 7);
+    assert_eq!(image_header_box.colourspace_unknown(), 0);
+    assert_eq!(image_header_box.intellectual_property(), 0);
+    assert_eq!(image_header_box.components_bits(), 16);
+    assert_eq!(image_header_box.values_are_signed(), false);
+
+    assert!(header_box.bits_per_component_box.is_none());
+
+    assert_eq!(header_box.colour_specification_boxes.len(), 1);
+    let colour_specification_box = header_box.colour_specification_boxes.first().unwrap();
+    assert_eq!(
+        colour_specification_box.method(),
+        Methods::EnumeratedColourSpace
+    );
+    assert_eq!(colour_specification_box.precedence(), 0);
+    assert_eq!(colour_specification_box.colourspace_approximation(), 0u8);
+    assert!(colour_specification_box.enumerated_colour_space().is_some());
+    assert_eq!(
+        colour_specification_box.enumerated_colour_space().unwrap(),
+        16u32
+    );
+
+    assert!(header_box.palette_box.is_none());
+
+    assert!(header_box.component_mapping_box.is_none());
+
+    assert!(header_box.channel_definition_box.is_none());
+
+    assert!(header_box.resolution_box.is_none());
+
+    assert_eq!(boxes.contiguous_codestreams_boxes().len(), 1);
+    let codestream_box = boxes.contiguous_codestreams_boxes().first().unwrap();
+    assert_eq!(codestream_box.length(), expected.file_length - 85);
+    assert_eq!(codestream_box.offset(), 85);
+
+    assert!(reader.seek(SeekFrom::Start(codestream_box.offset)).is_ok());
+    let codestream_decode_result = jpc::decode_jpc(&mut reader);
+    assert!(codestream_decode_result.is_ok());
+    let codestream = codestream_decode_result.unwrap();
```

Comment:
Nit, can we move the codestream tests/assertions to the jpc crate or the jpeg2000, as the intention is the jp2 crate only handles the decoding of the container. In theory we can have a codestream without the container, or a different container. I'm open to other views though.

Reply:
Possibly I can regenerate the test files without the JP2 headers and test in jpc. It would make for more data,  but they're fairly small.

Putting it in jp2000 feels wrong in that its mostly CLI.

Another option would be an integration test crate. Not great, but maybe the least layering violation.

Reply:
A separate integration test crate could work well and may be required later on anyway. I agree the CLI crate isn't ideal. What if we extract the codestream and place it in a file? Although there's no registered extension it seems .jpc/.j2k/.j2c is commonly used. This way we can minimise the codestream and respective tests in the jp2 crate. There may be a case for a command in the CLI to extract the codestream from a jp2 to make this easier (jpxml goes part of the way)

Reply:
OpenJPEG is happy to make codestream-only files. I'm currently experimenting with finer-grained test files, looks possible.

At some stage I think we'll need integration tests to cross-check values between the JP2 level boxes and the codestream segments. However for codestream parsing, its probably better to put it in jpc if we can avoid the jp2 dependency (or vice versa).

The extraction tool seems like a useful idea. Added that as #27